### ISSUE TITLE
Darkmode

### ISSUE THREAD

**Issue by jappi00** [pdoc3/pdoc#152]
Hey,

I think we all love Darkmodes. What about pdoc? I think a default integrated Darkmode would be great! What do you think about something like this? Thank you for this awesome package!

**Comment by kernc** [comment]
We do. :smile:

Integrated, but optional behind a [config toggle](https://pdoc3.github.io/pdoc/doc/pdoc/#custom-templates), would indeed make sense. Or even something purely JS-based that shows a light switch on the client side for the user to switch at will. :thinking: :bulb: 

I won't be able to work on this shortly, so further discussion / PR welcome!

**Comment by kernc** [comment]
Or, if the way of a config tunable, a dict of preferred style colors would empower the user more. Something like:
```py
# config.mako:

bright_color_theme = dict(foreground='black',
                          background='white',
                          link='...',
                          ...)
dark_color_theme = ...
color_theme = bright_color_theme  # Default
```

**Comment by kernc** [comment]
An even better idea is to provide a basic set of easily overridable colors in CSS variables (e.g. in _pdoc/templates/colors.css_) like it's currently for highlighting color:
https://github.com/pdoc3/pdoc/blob/cd84fb3b0c512a203742cfdfaa26ba6fb0e535aa/pdoc/templates/css.mako#L6-L8
https://github.com/pdoc3/pdoc/blob/cd84fb3b0c512a203742cfdfaa26ba6fb0e535aa/pdoc/templates/css.mako#L184-L186

**Comment by xcodz-dot** [comment]
Hello, I was also looking for dark theme for pdoc3, can you please tell what fields am i supposed to modify in css file?

**Comment by kernc** [comment]
@xcodz-dot it's not available yet. There's just [the idea](https://github.com/pdoc3/pdoc/issues/152#issuecomment-677903920) to separate colors hard-coded all over _css.mako_ into another file _colors.css_ (or maybe _theme.css_), with CSS variables defined on the _:root_ pseudo-class.

That new file then needs to be included in the _html.mako_ template:
https://github.com/pdoc3/pdoc/blob/aef1917a668ca9f3cd26373d1d69bcb5001108cc/pdoc/templates/html.mako#L389-L392

This sounds good enough in principle, but I think it's going to prove a **breaking change for anyone who's customized their _html.mako_** (but not _css.mako_). ~~Maybe a safer alternative would be to declare all those color variables already in _css.mako_ at the top, but then in _html.mako_ conditionally load _colors.css_ (if available) where interested users can put their individual color overrides. A standalone _colors.css_ could serve as a standalone color template.~~
Or maybe the best would be to declare color variables in a separate _colors.css_, and then import it **into _css.mako/mobile_ section**:
https://github.com/pdoc3/pdoc/blob/aef1917a668ca9f3cd26373d1d69bcb5001108cc/pdoc/templates/css.mako#L5-L8

In either case, someone needs to figure it out, and I'd appreciate a PR.

**Comment by xcodz-dot** [comment]
So the thing I have ended up with is to copy the default `css.mako` and now i am going to make some changes to it so it looks good and dark. If i succeed, i will also submit a PR for for seperate `css.mako` with commandline option changes for selecting inbuilt themes.

**Comment by xcodz-dot** [comment]
Ok, so looks like I have pretty much achieved a dark theme. I tested that theme on pdoc documentation. but there are somethings like warnings, todo, notes, etc... that i have to change color of, so is there something like `demo.py` so i can test my theme on it and verify all the aspects of it.

**Comment by kernc** [comment]
[`pdoc.test.example_pkg`](https://pdoc3.github.io/pdoc/doc/pdoc/test/example_pkg/#pdoc.test.example_pkg.reST_directives) tries out some features, including admonitions.

> i will also submit a PR for seperate `css.mako` with commandline option changes for selecting

I'd rather first have a solid mechanism for customizing the color theme. We can bundle some themes after the feature spreads and matures. Till then, having the user provide [`--template-dir`](https://pdoc3.github.io/pdoc/doc/pdoc/#custom-templates) pointing to their overrides is the preferred approach.

**Comment by xcodz-dot** [comment]
> [`pdoc.test.example_pkg`](https://pdoc3.github.io/pdoc/doc/pdoc/test/example_pkg/#pdoc.test.example_pkg.reST_directives) tries out some features, including admonitions.
> 
> > i will also submit a PR for seperate `css.mako` with commandline option changes for selecting
> 
> I'd rather first have a solid mechanism for customizing the color theme. We can bundle some themes after the feature spreads and matures. Till then, having the user provide [`--template-dir`](https://pdoc3.github.io/pdoc/doc/pdoc/#custom-templates) pointing to their overrides is the preferred approach.

that will be good if they use template-dir but what I meant is that `--theme` option could be added so the user could select inbuilt themes and the user can use `--template-dir` to override config of that inbuilt theme. basically `theme` option will also look for any packages installed that supply theme for pdoc ([plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/)).

In other words, `--theme` specifies inbuilt theme or python plugin theme and `--template-dir` specifies local overrides of the selected theme, by default `--theme` will be `light`. And as the collection grows more themes can be used with `--theme` option.

It's still up to you. I will still be happy to contribute my added themes.

Also, here is a alternative to `--theme`. We can format the input required by `--template-dir`, for example:

* Local theme: `docs/themes`
* Inbuilt or Plugin themes: `theme:dark`

Now in second case we can identify and resolve the path to the selected theme and I will be complete with dark theme by 3 days or most probably earlier.

**Comment by xcodz-dot** [comment]
Looks like my theme is complete and ready for use, also it's more like black theme so i will also try making a greyish dark theme

**Comment by jappi00** [comment]
> [`pdoc.test.example_pkg`](https://pdoc3.github.io/pdoc/doc/pdoc/test/example_pkg/#pdoc.test.example_pkg.reST_directives) tries out some features, including admonitions.
> 
> > i will also submit a PR for seperate `css.mako` with commandline option changes for selecting
> 
> I'd rather first have a solid mechanism for customizing the color theme. We can bundle some themes after the feature spreads and matures. Till then, having the user provide [`--template-dir`](https://pdoc3.github.io/pdoc/doc/pdoc/#custom-templates) pointing to their overrides is the preferred approach.

Do you have seen the Symbol/ swastika?

That is not funny.

**Comment by xcodz-dot** [comment]
> > [`pdoc.test.example_pkg`](https://pdoc3.github.io/pdoc/doc/pdoc/test/example_pkg/#pdoc.test.example_pkg.reST_directives) tries out some features, including admonitions.
> > > i will also submit a PR for seperate `css.mako` with commandline option changes for selecting
> > 
> > 
> > I'd rather first have a solid mechanism for customizing the color theme. We can bundle some themes after the feature spreads and matures. Till then, having the user provide [`--template-dir`](https://pdoc3.github.io/pdoc/doc/pdoc/#custom-templates) pointing to their overrides is the preferred approach.
> 
> Do you have seen the Symbol/ swastika?
> 
> That is not funny. It is in the footer.

what symbol are you talking about?

**Comment by kernc** [comment]
@jappi00 It's likewise not as grave as you're having it. See https://github.com/pdoc3/pdoc/issues/310. :sweat_smile: 

@xcodz-dot I'm not a huge fan of populous CLI switches that sooner or later nobody has proper count of (cf. `pip`, `pytest`). pdoc already uses a [config file](https://github.com/pdoc3/pdoc/blob/master/pdoc/templates/config.mako) to tune template-related stuff, and it works via `--template-dir` and `--config`, so that'd be a `theme=` config setting. That said, I'm not sure I want to bundle a bunch of themes at the moment. For a start, I'd like to see a robust, backwards-compatible, extensible mechanism for color/style overriding. Only then, if the feature spreads (determined e.g. via GitHub search), we can bundle a few styles based on popular opinion. Since color and style are nearly wholly bikeshedding territory, I think this is the best approach.

**Comment by xcodz-dot** [comment]
> @jappi00 It's likewise not as grave as you're having it. See #310. 😅
> 
> @xcodz-dot I'm not a huge fan of populous CLI switches that sooner or later nobody has proper count of (cf. `pip`, `pytest`). pdoc already uses a [config file](https://github.com/pdoc3/pdoc/blob/master/pdoc/templates/config.mako) to tune template-related stuff, and it works via `--template-dir` and `--config`, so that'd be a `theme=` config setting. That said, I'm not sure I want to bundle a bunch of themes at the moment. For a start, I'd like to see a robust, backwards-compatible, extensible mechanism for color/style overriding. Only then, if the feature spreads (determined e.g. via GitHub search), we can bundle a few styles based on popular opinion. Since color and style are nearly wholly bikeshedding territory, I think this is the best approach.

`theme=` will be great for me to add it up. And here is one more opinion because i see you really do not like to add up builtin themes and extra cli options. We can make use of [plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/) to extend themes. default for `theme=` will be `light`(inbuilt default), and we can use pypi package suffix `pdoc-theme`. Summarizing up: `theme=` option in config file; use of plugins for extending theme (I will submit my theme to pypi instead)

**Comment by xcodz-dot** [comment]
By the way here is the almost final theme:

<img width="960" alt="s3" src="https://user-images.githubusercontent.com/71920621/109264926-bc1ee400-782b-11eb-855a-c046bc0035a0.PNG">

I changed scroll bar color and changed the color of code between back ticks. Also the location of the theme that i am working on is [here](https://github.com/xcodz-dot/denver-api/tree/master/pdoc)

**Comment by xcodz-dot** [comment]
> Do you have seen the Symbol/ swastika?

looks like symbols were wiped out from my locally generated docs

**Comment by jappi00** [comment]
> > Do you have seen the Symbol/ swastika?
> 
> looks like symbols were wiped out from my locally generated docs

Ahh, sry for that. I‘am from germany. We are overreacting when we see things like this.

**Comment by xcodz-dot** [comment]
so my last word is that i am gonna implement plugin system into pdoc

**Comment by kernc** [comment]
Indeed, plugin support would be a heavy and questionably-worthwhile undertaking. At this stage still, extracting CSS colors into a separate overridable _colors.css_, [as above](https://github.com/pdoc3/pdoc/issues/152#issuecomment-785275318), seems the most reasonable course of action.

**Comment by xcodz-dot** [comment]
> Indeed, plugin support would be a heavy and questionably-worthwhile undertaking. At this stage still, extracting CSS colors into a separate overridable _colors.css_, [as above](https://github.com/pdoc3/pdoc/issues/152#issuecomment-785275318), seems the most reasonable course of action.

and yes that's what i did. so i guess i will soon host the repo for the dark theme. I have some work to do currently

**Comment by MikhailZex** [comment]
Well here is another dark mode [css.mako](https://gist.github.com/MikhailZex/60525d37e7eb06cef903994fce4dc2e7).

I do also have a suggestion: create a new repo `pdoc3/pdoc-user-themes` where pdoc users can submit their own themes. That way people can both get the themes they want while you do not have to make any changes to the existing code. Win-Win.

**Comment by pm5k** [comment]
From what I'm reading, it seems to me like this needs a little bit of fleshing out requirements-wise. I am tempted to give it a whack, but even if someone else decides to, might be nice to have these things listed in one space.

## Requirements
* **Support for dark mode**, [read more](https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web) as this article does a pretty good job at explaining methods, tips and considerations for dark mode implementations.
* **Changes to be non-breaking**, users who update to a release containing this code should not have their layouts break / look funny.
* **Override / Theming support** - users should be free to modify CSS sources or override them to easily make their own custom themes for the entirety of the docs' "look and feel".

## Considerations

### Dark Mode support:
It seems like currently the CSS files are split up by "platform", I.E. mobile, desktop and print (although print is the odd one out here). I (personally) feel like there is no need for this at all, and the base style should just be one CSS file utilising an import of CSS vars (either from a preceding stylesheet or from the same one). I would also carefully re-group component-scope declarations in order to position them for the inevitable:

```css
    :root {
        --bg-primary: #fff;
        --fc-primary: #000;
    }

    @media (prefers-color-scheme: dark) {
        :root {
            --bg-primary: #000;
            --fc-primary: #fff;
        }
    }

    body {
        background-color: var(--bg-primary);
        color: var(--fc-primary);
    }
```

What's great about this approach is that we can leverage `mako` to spit out these vars so we don't have to duplicate ourselves in writing them for `:root` and `"dark-mode" :root`, but more importantly it would make it incredibly easy to theme everything as well in a standardised way. For instance, a naive approach could be below. Do excuse the tab and newline mix in the code, I was hoping to see if this mattered (it did not), so the end result would be much cleaner and thankfully - easier to read.

`css.mako`
```mako
<%def name="make_colors()">
  <%
    # Name, default, dark mode
    css_colors = [
      ("--bg-primary", "#FFF", "#000"),
      ("--bc-primary", "#000", "#FFF")
    ]
    default = [f"\n\t{c[0]}: {c[1]};" for c in css_colors]
    dark = [f"\n\t\t{c[0]}: {c[2]};" for c in css_colors]

    vardef = (
      f"\n:root {{{ ''.join(default) }\n}}"
      + f"\n@media (prefers-color-scheme: dark) {{"
      + f"\n\t:root {{{ ''.join(dark) }\n}}"
      + f"\n}}"
    )
  %>
  ${ vardef }
</%def>

## Minify removed for output testing..
<%def name="desktop()">
  ${make_colors()}
...
```

`Output`
```html
<style media="screen and (min-width: 700px)">
:root {
--bg-primary: #FFF;
--bc-primary: #000;
}
@media (prefers-color-scheme: dark) {
:root {
--bg-primary: #000;
--bc-primary: #FFF;
}
}
@media screen and (min-width: 700px) {
...
```

This means that these color variables could be config(or file)-driven allowing drop-in replacements for the sake of theming. Or you could have override files which Mako could diff between them and the default config to only apply variables that are included and leave un-overridden ones as default.

I think this could be a clean approach, albeit it would require a small rewrite to the `css.mako` file.

### Non-breaking changes
Honestly, this it tough, because frankly if people are editing the `css.mako` file directly, there is little you can do to ensure continuous function of styles for such a change. In this scenario you're looking at a pre-release announcement with carefully detailed migration instructions for custom css.

For those who did not touch the css source, nothing changes. Any variations introduced in-html are frankly outside that context and are brittle by default. Say the use of inline styles - it's neither recommended (unless done via some automated and robust means) nor is it something that fosters compatibility. It's not necessarily easy for me to suggest people "suck it up" in this case, so maybe a note around moving custom components and styles out of templates and into the CSS file would be enough of a nudge.

### Theming
With the change proposed/detailed above, I believe theming would become trivial. In-fact you could even provide a small pre-rendered "generator" SPA which would give users a nice UI around color changes which reflect how they will appear in the UI. Sort of like [Bootstrap Magic](https://pikock.github.io/bootstrap-magic/app/index.html#!/editor), but incredibly scaled down for the purposes of function over form and small bundle size. It's just an idea, but would greatly extend support for theming.


## In closing
I am unsure if I have missed any obvious points, but so far from what I have looked at, this could be a nice way to introduce dark mode and theming to PDOC, and hopefully not mess stuff up too much retroactively.

**Comment by pm5k** [comment]
An update from my side - I have got a preliminary working dark mode and easier theming support (See gifs below for demo of this).

I retained all of the original styling of `pdoc` with the exception of content width, as it was wasting too much space by not filling all the space to the right of itself, and also the media queries as they were slightly cleaned up and the mobile/tablet/portable query was changed to breakpoint `1024` as this is what is defined as standard now for most responsive designs. This can be reverted if needed, however.

The color info was moved out of `css.mako`, but all the other css styles were retained.

### Changes
The way I've implemented this is that it is driven via `config.mako`. The first change is that inside of this file you will see:

`config.mako`
```mako
## ...
hljs_style_tuple = ('github', 'atom-one-dark')  ## Replaces `hljs_style`
## ...
css_theme = 'default'
```

The `hljs_style_tuple` is there to support on-the-fly code highlighting swaps, so that if the user's system changes color mode preference or the user toggles this themselves, the documentation will adapt and not only swap between dark/light but also apply the same swap to HLJS. Therefore this value now has to have one style definition for light mode and one for dark mode.

The `css_theme` is literally there to tell mako which theme template to load. Each template has its own dark mode included by design. The `default` theme for instance is the vanilla `pdoc3` theme when light mode is enabled and when switched to dark mode it adopts the dark theme of github's docs.

The template structure has changed thusly:
```bash
templates/
|__themes/
|     |__custom.mako
|     |__default.mako
|_ config.mako
|_ ...
```
Each theme is it's own file and contains the following structure:

`default.mako`
```mako
<%!
    from pdoc.html_helpers import minify_css
    colors = [
        ("--bg-highlight-color", "#FFEE99", "#3E4951"),
        ("--bg-default", "#FFFFFF", "#22282D"),
        ## ...
    ]

    default = [f"{color[0]}: {color[1]};" for color in colors]
    dark = [f"{color[0]}: {color[2]};" for color in colors]

    root_light = f":root[data-theme='theme-light'] {{{''.join(default)}}}"
    root_dark = f":root[data-theme='theme-dark'] {{{''.join(dark)}}}"
    root_light_auto = f"@media (prefers-color-scheme: light) {{ :root[data-theme='theme-light'] {{{''.join(default)}}} }}"
    root_dark_auto = f"@media (prefers-color-scheme: dark) {{ :root[data-theme='theme-dark'] {{{''.join(dark)}}} }}"

    color_variables = f"{root_light}{root_dark}{root_light_auto}{root_dark_auto}"
%>

<%def name="style()" filter="minify_css">
    ${color_variables}
</%def>
```

The css template also changed to now contain styling sans the color values and instead relies on the provided vars from the theme file. Which keeps the structure completely intact, but allows easy theming. All someone has to do is copy the `default.mako` and rename it, then change the color values, or fill in `custom.mako` to keep it as a private project theme.

## Left to implement

- Retain theme pref in localStorage.
- Clean up the code.

## Demo
![](https://media.giphy.com/media/dR961X5zjjxlUhB1xE/giphy.gif)
![](https://media.giphy.com/media/X4uxoIg7rWAME9LEr3/giphy.gif)

**Comment by Mylloon** [comment]
Hope your work will be merged soon 🙏

**Comment by rolandsfr** [comment]
Anyone has come up with a solid solution yet?

**Comment by tboerstad** [comment]
@pm5k Is dark mode still something that might show up? It looks great

**Comment by amaank404** [comment]
As far as I am concerned, I do not think this is happening anytime soon, you can use my theme [here](https://github.com/amaank404/denver-api/tree/master/pdoc) if you would like to. You can copy the folder to where you run the commands and while generating the docs with the pdoc command just add this flag: `--template-dir pdoc`. I am not currently doing anything python related so this is all the help I can provide. For a preview look at [this](https://github.com/pdoc3/pdoc/issues/152#issuecomment-786450437)
